### PR TITLE
Support Math keyword in native_functions.yaml.

### PR DESCRIPTION
**Pull request by ailzhang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44993 Define macro for DispatchKey::toString.
* #44933 Align casing in test_dispatch with dispatch keys.
* **#44556 Support Math keyword in native_functions.yaml.**
* #44354 Add alias dispatch key Math.

Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

### CODE REVIEW

File: aten/src/ATen/native/README.md [hunk 1]
```diff
@@ -277,6 +277,18 @@ them the same thing!)
 If two backends have the same dispatch function, you can write `CPU, CUDA: func`
 to reuse the same function name in both cases.
 
+TODO: expand this section.
+
+Available backend options can be found at
+https://github.com/pytorch/pytorch/blob/master/tools/codegen/gen.py#L970.
+In addition to backends above, we also support keyword `Math` which is an alias
+that maps to all backends. In other words, function registered to `Math` key
+should be a plain mathematical composition of other `at::` functions and works for any backend.
```
Comment (zou3519) [inline aten/src/ATen/native/README.md:286]: (drive-by comment): Are things registered to the Math key required to be functional (e.g., no in-place operators), or are in-place operations OK?

File: aten/src/ATen/native/README.md [hunk 2]
```diff
@@ -277,6 +277,18 @@ them the same thing!)
 If two backends have the same dispatch function, you can write `CPU, CUDA: func`
 to reuse the same function name in both cases.
 
+TODO: expand this section.
```
Comment (zou3519) [inline aten/src/ATen/native/README.md:280]: (another drive-by comment): I don't know what the Math key is but it looks interesting. Would love to see an explanation of what it's used for here

File: aten/src/ATen/native/README.md (hunk 1 above)
Comment (ailzhang) [reply aten/src/ATen/native/README.md:286]: It doesn't have to be functional. I'd think of `Math` key being used to store the old `composite` implementations. 
Imagine we used to implement `native_group_norm` using `at::native_batch_norm` (it doesn't need to be functional), once we add an optimized kernel for certain backend, say
 ```
func: native_group_norm
CPU: group_norm_cpu
CUDA: group_norm_cuda
# Math: old_native_group_norm
```
`Math` is a way to tell all other backends you can still use `old_native_group_norm` without issues. Before this PR, once we add `CPU: group_norm_cpu`, backends like XLA are suddenly required to register a kernel to `native_group_norm` otherwise there's no kernel available for that key. 
Hope this explains it a little better, let me know if you have followup questions!

File: aten/src/ATen/native/README.md (hunk 2 above)
Comment (ailzhang) [reply aten/src/ATen/native/README.md:280]: https://fb.quip.com/jiYeAa58PDX0 explains it pretty well. But yea I should improve docs here as well. ;)

File: tools/autograd/gen_variable_type.py [hunk 3]
```diff
@@ -370,7 +370,8 @@
 
 # Generate a file that lists all functions and their schema string. Used for XLA
 REGISTRATION_DECLARATION = CodeTemplate("""\
-${return_type} ${api_name}(${declaration_formals}); // {"schema": "${schema_string}", "compound": "${compound}"}
+${return_type} ${api_name}(${declaration_formals}); \
+// {"schema": "${schema_string}", "compound": "${compound}", "has_math_kernel": "${has_math_kernel}"}
```
Comment (ezyang) [inline tools/autograd/gen_variable_type.py:374]: Wondering if we can just port this codegen straight to the new codegen, then we don't have to expose `has_math_kernel` in Declarations.yaml.

File: tools/codegen/gen.py [hunk 4]
```diff
@@ -1016,16 +1017,30 @@ def make_file_manager(install_dir: str) -> FileManager:
...
+            compute_type_method('Math', target=Target.DEFINITION, op_registration_whitelist=op_registration_whitelist),
             native_functions)),
-        'function_registrations': list(mapMaybe(
+
+        'function_registrations':
+        list(mapMaybe(
             compute_type_method(None, target=Target.REGISTRATION, op_registration_whitelist=op_registration_whitelist),
+            native_functions)) +
+        list(mapMaybe(
+            compute_type_method('Math', target=Target.REGISTRATION, op_registration_whitelist=op_registration_whitelist),
```
Comment (ezyang) [inline tools/codegen/gen.py:1043]: This happens to work but the reason is not so good. When you dunk this in `function_registrations` it gets dumped in:

```
TORCH_LIBRARY(aten, m) {
  ${function_registrations};
```

This doesn't specify a dispatch key specifically (unlike the CPU/CUDA files). Everything works out because we always generate `torch::dispatch`:

```
                # Annotate it with dispatch information if necessary
                #
                # NB: In the ordinary, TypeDerived code generation work flow, specification
                # of the backend is handled by the enclosing block, so the torch::dispatch
                # invocation here is strictly unnecessary.  However, in the fbcode mobile
                # only workflow using per-op registration, these registrations will get dumped
                # in a TORCH_LIBRARY_FRAGMENT that does not have an ambient backend.  So
                # the torch::dispatch specification here is important!  See
                # Note [Redundancy in registration code is OK] for how we handle redundant info.
                if dispatch is not None:
                    payload = f"torch::dispatch(DispatchKey::{dispatch},\n{payload})\n"
```

With this change, this comment is no longer true, since you're also relying on the torch::dispatch call to make sure Math goes to the right place. But it seems to me that it would be clearer if you just made a separate `math_registrations` entry and then put it in a dedicated `Math` registration block, at which point the comment above would remain true.

File: tools/autograd/gen_variable_type.py [hunk 5]
```diff
@@ -654,12 +655,12 @@ def gen_variable_type(out, aten_declarations, template_path):
             registration_declarations.append(
                 REGISTRATION_DECLARATION.substitute(declaration,
                                                     declaration_formals=declaration_formals,
-                                                    compound='false'))
+                                                    compound='False'))
```
Comment (ezyang) [inline tools/autograd/gen_variable_type.py:658]: Not JSON anymore? lol

File: tools/autograd/gen_variable_type.py (hunk 3 above)
Comment (ailzhang) [reply tools/autograd/gen_variable_type.py:374]: Unfortunately it has to stay in autograd codegen for now since we need access to `whether an op has derivative` to cover DispatchSub ops. Whether the kernel comes from disptach stub is not available anywhere in codegen and can only be approximated by the fact that almost all dispatch stub ops have derivatives). So we need to add this information in native_functions.yaml before we could move this codegen. More details can be found at: https://fb.quip.com/1B61AfNcuozU

Comment (ezyang) [reply tools/autograd/gen_variable_type.py:374]: OK. We'll be out of this situation soon, but this is fine for this PR.

File: aten/src/ATen/native/README.md (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/native/README.md:280]: Yeah, let's publish the design on a GitHub issue so non-FB people can see it.

File: aten/src/ATen/native/group_norm.cpp [hunk 6]
```diff
@@ -133,5 +133,29 @@ Tensor group_norm(
 DEFINE_DISPATCH(GroupNormKernel);
 DEFINE_DISPATCH(GroupNormBackwardKernel);
 
+std::tuple<at::Tensor, at::Tensor, at::Tensor> math_group_norm(
```
Comment (ezyang) [inline aten/src/ATen/native/group_norm.cpp:136]: For CR purposes it would be helpful to cite where this is coming from

File: aten/src/ATen/test/math_kernel_test.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+using namespace at;
+
+#define ASSERT_ALLCLOSE_TOLERANCES(t1, t2, atol, rtol) \
+  ASSERT_TRUE(t1.is_same_size(t2));                    \
+  ASSERT_TRUE(t1.allclose(t2, atol, rtol));
+
+// Ideally we want to test both forward and backward on math kernels but I
+// haven't found an easy way to do it.  Currently we only test forward here
+// and rely on backward tests of each at:: function used in math kernels.
```
Comment (ezyang) [inline aten/src/ATen/test/math_kernel_test.cpp:13]: This is definitely something we should work on making easier, though I'm not exactly sure how. Perhaps APIs for supporting static runtime would make it easier, as they would make it possible to ask for a specific kernel.

Absent machinery like this, I think the easiest way to do this is to just make the math kernel available at another name (i.e., make a new `native_functions.yaml` entry; and then just do the testing from Python. Maybe @mruberry has some other ideas.

File: aten/src/ATen/native/group_norm.cpp [hunk 8]
```diff
@@ -133,5 +133,29 @@ Tensor group_norm(
...
+  if (weight.defined() && bias.defined()) {
+    out = bias.view(affine_param_shape).addcmul(out, weight.view(affine_param_shape), 1);
+  } else if (weight.defined()) {
+    out = out.mul(weight.view(affine_param_shape));
+  } else if (bias.defined()) {
+    out = out.add(bias.view(affine_param_shape));
+  }
+  at::Tensor mean = std::get<1>(outputs).view({N, group});
+  at::Tensor rstd = std::get<2>(outputs).view({N, group});
+  return std::make_tuple(out, mean, rstd);
```
Comment (ezyang) [inline aten/src/ATen/native/group_norm.cpp:158]: I'm going to trust that this was lifted directly from XLA and has already been code reviewed.

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/README.md [hunk 9]
```diff
@@ -277,6 +277,18 @@ them the same thing!)
...
+
+Available backend options can be found at
+https://github.com/pytorch/pytorch/blob/master/tools/codegen/gen.py#L970.
+In addition to backends above, we also support keyword `Math` which is an alias
+that maps to all backends. In other words, function registered to `Math` key
```
Comment (bhosmer) [inline aten/src/ATen/native/README.md:285]: Do we want to note here that `Math` maps to all autograd keys as well? Or is this doc expressly written to avoid pulling in a discussion of autograd?

File: aten/src/ATen/native/README.md [hunk 10]
```diff
@@ -277,6 +277,18 @@ them the same thing!)
...
+should be a plain mathematical composition of other `at::` functions and works for any backend.
+
+If you add `dispatch` section to any API that didn't have it before, you **have to** move
+the old implementation to `Math` field so that it's still available for other backends to use.
```
Comment (bhosmer) [inline aten/src/ATen/native/README.md:289]: We can also enforce this, right? (Maybe it's further down the PR...)

**Review by bhosmer — APPROVED** [review]

File: aten/src/ATen/native/group_norm.cpp (hunk 6 above)
Comment (ailzhang) [reply aten/src/ATen/native/group_norm.cpp:136]: I wrote the pytorch/xla version based on the old pytorch composite implementation. So it should be fine here ;)

File: aten/src/ATen/native/README.md (hunk 10 above)
Comment (ailzhang) [reply aten/src/ATen/native/README.md:289]: Good point! We currently cannot enforce it easily since we don't track whether ops has dispatch sections and whether they change. But this should be possible in the terminal state where we can track this through our API provided to backend extensions. ;) So I'd save this as a followup task. ;)

Comment (bhosmer) [reply aten/src/ATen/native/README.md:287]: Ah yeah - actually I oversimplifying , picturing a world where we could simply insist that if you had any explicit dispatch entries at all, then `Math` had to be one of them (say, by adding a semantic check in `NativeFunction.from_yaml`). But of course we'd need a codemod (or something) before our current declarations would fail that test, so yeah - followup task sounds right!

**Review by bhosmer — APPROVED** [review]
Looks good!

File: aten/src/ATen/test/math_kernel_test.cpp (hunk 7 above)
Comment (mruberry) [reply aten/src/ATen/test/math_kernel_test.cpp:13]: Without too much context I'll also vote for (re)using Python.
